### PR TITLE
Scheduler constraints

### PR DESCRIPTION
**Pull request by anwar6953**
Implement TimeRangeConstraint and DelayConstraint.

### CODE REVIEW

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/ScheduleConfigurer.java [hunk 1]
```diff
@@ -32,9 +32,9 @@
 
   ScheduleConfigurer delayRun(long delayMillis);
 
-  ScheduleConfigurer setTimeRange(int startHour, int endHour);
+  ScheduleConfigurer setTimeRange(String startTime, String endTime);
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/schedule/ScheduleConfigurer.java:35]: It's a bit unclear what's the content of the string. Will there be some doc?

File: cdap-api/src/main/java/co/cask/cdap/internal/schedule/constraint/Constraint.java [hunk 2]
```diff
@@ -19,5 +19,5 @@
 /**
  * A constraint that is checked before executing a schedule.
  */
-public abstract class Constraint {
+public interface Constraint {
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/internal/schedule/constraint/Constraint.java]: Why this change?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/AbstractConstraint.java [hunk 3]
```diff
@@ -24,13 +24,15 @@
 /**
  * A constraint that is checked before executing a schedule.
  */
-public class ConstraintChecker {
+public abstract class AbstractConstraint implements Constraint {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/AbstractConstraint.java]: I think it's better to have an interface `CheckableConstraint` or `JobConstraint` that is internal to the scheduler in app-fabric and have the abstract class implementing that interface.

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/ScheduleConfigurer.java (hunk 1 above)
Comment (anwar6953) [reply cdap-api/src/main/java/co/cask/cdap/api/schedule/ScheduleConfigurer.java:35]: Yea, there's already a JIRA for that: https://issues.cask.co/browse/CDAP-11359

File: cdap-api/src/main/java/co/cask/cdap/internal/schedule/constraint/Constraint.java (hunk 2 above)
Comment (anwar6953) [reply cdap-api/src/main/java/co/cask/cdap/internal/schedule/constraint/Constraint.java]: the implementation of the Constraints need to extend from a class in `cdap-app-fabric`, the `AbstractConstraint`, so had to make this one just an interface.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/AbstractConstraint.java (hunk 3 above)
Comment (anwar6953) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/AbstractConstraint.java]: Why isn't it sufficient that this `AbstractConstraint` acts as the parent of all constraint implementation classes?

Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/AbstractConstraint.java]: In general, interface is for declaring capability, while class is for hierarchy. E.g. you can have `abstract class Constraint` in API, which represents hierarchy of constraints. Some Constraints (although all in this particular case) is "checkable".

Comment (anwar6953) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/AbstractConstraint.java]: Is something like this what you're recommending?
`DelayConstraint` ==> `AbstractConstraint` ==> `JobConstraint` ==> `Constraint`

`DelayConstraint` is a concrete class in cdap-app-fabric.
`AbstractConstraint` is an abstract class in cdap-app-fabric.
`JobConstraint` is an interface in cdap-app-fabric.
`Constraint` is an interface in cdap-api.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/AbstractCheckableConstraint.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+package co.cask.cdap.internal.app.runtime.schedule.constraint;
+
+import co.cask.cdap.internal.app.runtime.schedule.ProgramSchedule;
+import co.cask.cdap.internal.schedule.constraint.Constraint;
+
+/**
+ * A constraint that is checked before executing a schedule.
+ */
+public abstract class AbstractCheckableConstraint extends Constraint implements CheckableConstraint {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/AbstractCheckableConstraint.java]: Seems like you actually don't need this class since there is no implementation?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/CheckableConstraint.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+package co.cask.cdap.internal.app.runtime.schedule.constraint;
+
+import co.cask.cdap.internal.app.runtime.schedule.ProgramSchedule;
+
+/**
+ * A constraint that can be checked for satisfaction.
+ */
+public interface CheckableConstraint {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/CheckableConstraint.java]: hum.. now this becomes interesting, since this interface doesn't not extend from `Constraint` (which it can't).

Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/CheckableConstraint.java]: Maybe for simplicity for now, just have this:

```java
// This in cdap-api
public abstract class Constraint {
}

// This in app-fabric
public abstract class CheckableConstraint extends Constraint {
  public abstract CheckResult check(ProgramSchedule schedule, ConstraintContext context);
  
  public static final class CheckResult {
    ...
  } 
}

// This is an implementation in app-fabric
public final class TimeRangeConstraint extends CheckableConstraint {
  ...
}

// This is the usage in constraint checker
for (Constraint constraint : schedule.getConstraints()) {
  if (constraint instanceof CheckableConstraint) {
    ...
  }
}
```

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/AbstractCheckableConstraint.java (hunk 4 above)
Comment (anwar6953) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/AbstractCheckableConstraint.java]: This helps so that every implementation doesn't need to extend `Constraint` and implement `CheckableConstraint`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/DelayConstraint.java [hunk 6]
```diff
@@ -16,16 +16,33 @@
...
   private final long millisAfterTrigger;
 
   public DelayConstraint(long millisAfterTrigger) {
+    Preconditions.checkArgument(millisAfterTrigger >= 0,
+                                "Configured delay '%s' must not be a negative value.", millisAfterTrigger);
     this.millisAfterTrigger = millisAfterTrigger;
   }
+
+  public ConstraintResult check(ProgramSchedule schedule, ConstraintContext context) {
+    return check(System.currentTimeMillis(), context.getJob().getCreationTime());
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/DelayConstraint.java]: Better have the "checkTime" exposed from the `ConstraintContext` and have constraint check to use that instead of `System.currentTimeMillis`. This will make testing easier and we can provide the same check time for all constraints in the same job for consistency.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/CheckableConstraint.java (hunk 5 above)
Comment (anwar6953) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/CheckableConstraint.java]: :+1:
Removed the interface in cdap-app-fabric.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/DelayConstraint.java [hunk 7]
```diff
@@ -16,16 +16,33 @@
...
+  }
+
+  @VisibleForTesting
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/DelayConstraint.java]: This is exactly what I meant by making test easier above. Introducing this extra method just for testing making this class less readable and easier to make mistake in future (e.g. some extra logic added in the `check` method above may not get tested appropriately).

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/notification/Notifications.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+import com.google.gson.GsonBuilder;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Utility class for interacting with {@link Notification}s.
+ */
+public final class Notifications {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/notification/Notifications.java]: Do we need to introduce this util class?? Can the `forPartitions` be just static factory method in `Notification` class?

File: cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java [hunk 9]
```diff
@@ -284,12 +287,37 @@ private boolean runReadyJobs() throws IOException, DatasetManagementException {
...
     private boolean constraintsSatisfied(Job job) {
-      // TODO: implement constraint checking
+      for (Constraint constraint : job.getSchedule().getConstraints()) {
+        if (!(constraint instanceof AbstractCheckableConstraint)) {
+          // this shouldn't happen, since all Constraint implementations should extend AbstractConstraint
+          throw new IllegalArgumentException("Implementation of Constraint must extend AbstractConstraint");
+        }
+        AbstractCheckableConstraint abstractConstraint = (AbstractCheckableConstraint) constraint;
+        ConstraintResult result = abstractConstraint.check(job.getSchedule(), new ConstraintContext(job));
+        if (result != ConstraintResult.SATISFIED) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java]: I thought in the very beginning we agreed that the `ProgramSchedule` would only carry one `Constraint` field, which it will be the constraint to check or a `AndConstraint` which "and" together multiple constraints. This allow future enhancement much easier.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/notification/Notifications.java (hunk 8 above)
Comment (anwar6953) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/notification/Notifications.java]: moved.

**Comment by chtyim** [comment]
I was wrong about the `List<Constraint>` comment, so just keep it as a List.

**Comment by anwar6953** [comment]
The other comments are addressed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/TimeRangeConstraint.java [hunk 10]
```diff
@@ -16,24 +16,73 @@
...
+      endHour = calendar.get(Calendar.HOUR_OF_DAY);
+      endMinute = calendar.get(Calendar.MINUTE);
+
+      Preconditions.checkArgument(startDate.compareTo(endDate) < 0, "The start time must be before the end time.");
+    } catch (ParseException e) {
+      throw new IllegalArgumentException("Failed to parse time.", e);
+    }
+  }
+
+  public ConstraintResult check(ProgramSchedule schedule, ConstraintContext context) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/TimeRangeConstraint.java]: Miss the `@Override`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/DurationSinceLastRunConstraint.java [hunk 11]
```diff
@@ -16,16 +16,20 @@
...
-public class DurationSinceLastRunConstraint extends Constraint {
+public class DurationSinceLastRunConstraint extends AbstractCheckableConstraint {
 
   private final long millisSinceLastRun;
 
   public DurationSinceLastRunConstraint(long millisSinceLastRun) {
     this.millisSinceLastRun = millisSinceLastRun;
   }
+
+  public ConstraintResult check(ProgramSchedule schedule, ConstraintContext context) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/DurationSinceLastRunConstraint.java]: Miss `@Override`

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/ConcurrencyConstraint.java [hunk 12]
```diff
@@ -16,16 +16,20 @@
...
-public class ConcurrencyConstraint extends Constraint {
+public class ConcurrencyConstraint extends AbstractCheckableConstraint {
 
   private final int maxConcurrency;
 
   public ConcurrencyConstraint(int maxConcurrency) {
     this.maxConcurrency = maxConcurrency;
   }
+
+  public ConstraintResult check(ProgramSchedule schedule, ConstraintContext context) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/ConcurrencyConstraint.java]: Miss `@Override`

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/DelayConstraint.java [hunk 13]
```diff
@@ -16,16 +16,28 @@
...
 
   private final long millisAfterTrigger;
 
   public DelayConstraint(long millisAfterTrigger) {
+    Preconditions.checkArgument(millisAfterTrigger >= 0,
+                                "Configured delay '%s' must not be a negative value.", millisAfterTrigger);
     this.millisAfterTrigger = millisAfterTrigger;
   }
+
+  public ConstraintResult check(ProgramSchedule schedule, ConstraintContext context) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/DelayConstraint.java]: Miss `@Override`

File: cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java [hunk 14]
```diff
@@ -284,12 +287,38 @@ private boolean runReadyJobs() throws IOException, DatasetManagementException {
...
     private boolean constraintsSatisfied(Job job) {
-      // TODO: implement constraint checking
+      for (Constraint constraint : job.getSchedule().getConstraints()) {
+        if (!(constraint instanceof AbstractCheckableConstraint)) {
+          // this shouldn't happen, since all Constraint implementations should extend AbstractConstraint
+          throw new IllegalArgumentException("Implementation of Constraint must extend AbstractConstraint");
+        }
+        AbstractCheckableConstraint abstractConstraint = (AbstractCheckableConstraint) constraint;
+        ConstraintResult result = abstractConstraint.check(job.getSchedule(),
+                                                           new ConstraintContext(job, System.currentTimeMillis()));
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java]: Should be using the same `context` instance for the same job?

Comment (anwar6953) [reply cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java]: and just modify the `checkTime` on the same context object?

**Comment by chtyim** [comment]
Just couple comments left.

File: cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java (hunk 14 above)
Comment (anwar6953) [reply cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java]: nevermind, I get what you mean. updated.

**Comment by chtyim** [comment]
LGTM

**Comment by anwar6953** [comment]
This PR adds two new constraints; the test cases added for those constraints are passing.
Other test cases are failing on the integration branch, and these changes don't affect those test cases.
Merging.
